spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java

```diff
@@ -0,0 +1,334 @@
+			EmbeddedHeadersMessageConverter();
+
+	/**
+	 * Indicates whether the implementation and the message broker have
+	 * native support for message headers. If false, headers will be
+	 * embedded in the message payloads.
+	 */
+	private final boolean supportsHeadersNatively;
+
+	/**
+	 * Indicates what headers are to be embedded in the payload if
+	 * {@link #supportsHeadersNatively} is true.
+	 */
+	private final String[] headersToEmbed;
+
+	public AbstractMessageChannelBinder(boolean supportsHeadersNatively, String[] headersToEmbed) {
+		this.supportsHeadersNatively = supportsHeadersNatively;
+		this.headersToEmbed = headersToEmbed;
+	}
+
+	/**
+	 * Binds an outbound channel to a given destination. The implementation delegates to
+	 * {@link #createProducerDestinationIfNecessary(String, ProducerProperties)}
+	 * and {@link #createProducerMessageHandler(String, ProducerProperties)} for
+	 * handling the middleware specific logic.
+	 * @param destination        the name of the destination
+	 * @param outputChannel      the channel to be bound
+	 * @param producerProperties the {@link ProducerProperties} of the binding
+	 * @return the Binding for the channel
+	 */
+	@Override
+	public final Binding<MessageChannel> doBindProducer(final String destination, MessageChannel outputChannel,
+			final P producerProperties) {
+		Assert.isInstanceOf(SubscribableChannel.class, outputChannel,
+				"Binding is supported only for SubscribableChannel instances");
+		createProducerDestinationIfNecessary(destination, producerProperties);
+		final MessageHandler producerMessageHandler;
+		try {
+			producerMessageHandler = createProducerMessageHandler(destination, producerProperties);
+		}
+		catch (Exception e) {
+			if (e instanceof BinderException) {
+				throw (BinderException) e;
+			}
+			else {
+				throw new BinderException("Exception thrown while building outbound endpoint", e);
+			}
+		}
+
+		((SubscribableChannel) outputChannel).subscribe(
+				new SendingHandler(producerMessageHandler, !this.supportsHeadersNatively && HeaderMode.embeddedHeaders
+						.equals(producerProperties.getHeaderMode()), this.headersToEmbed));
+
+		return new DefaultBinding<MessageChannel>(destination, null, outputChannel,
+				producerMessageHandler instanceof Lifecycle ? (Lifecycle) producerMessageHandler : null) {
+
+			@Override
+			public void afterUnbind() {
+				afterUnbindProducer(destination, producerProperties);
+			}
+		};
+	}
+
+	/**
+	 * Creates target destinations for outbound channels. The implementation is middleware-specific.
+	 * @param name the name of the producer destination
+	 * @param properties producer properties
+	 */
+	protected abstract void createProducerDestinationIfNecessary(String name, P properties);
+
+	/**
+	 * Creates a {@link MessageHandler} with the ability to send data to the
+	 * target middleware. If the returned instance is also a {@link Lifecycle},
+	 * it must be started by the method and it will be stopped automatically
```

Comment:
what's the rationale for requiring it to be started before returned by the implementation, rather than starting it from the common superclass IF it's `Lifecycle` (could even check `isRunning()` so that it's optional)